File: leetcode/Python/0560-m-subarray-sum-equals-k.py
"""
Time: O(n)
Space: O(n)
"""
import collections
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        d = collections.defaultdict(int)
        d[0] = 1
        prefixSum = 0
        res = 0
        for i in range(len(nums)):
            prefixSum += nums[i]
            if prefixSum - k in d:
                res += d[prefixSum - k]
            d[prefixSum] += 1
        return res

File: leetcode/C++/0560-m-subarray-sum-equals-k.cpp
// Prefix sum
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        if (nums.empty()) return 0;
        unordered_map<int, int> counts{{0, 1}};
        int res = 0, sum = 0;
        for (int num : nums) {
            sum += num;
            // cur_prefix_sum - some_prefix_sum = k
            if (counts.find(sum - k) != counts.end())
                res += counts[sum - k];
            counts[sum]++;
        }
        return res;
    }
};